File: catkin_ws/src/learning_service/src/turtle_spawn.cpp
/**
* send /spawn service, service data type is turtlesim::Spawn
*/
#include <ros/ros.h>
#include <turtlesim/Spawn.h>

int main(int argc, char** argv)
{
	// init ROS node
	ros::init(argc, argv, "turtle_spawn");
	
	// create node handler
	ros::NodeHandle node;

	// when find /spawn service, create a service client
	ros::service::waitForService("/spawn");
	ros::ServiceClient add_turtle = node.serviceClient<turtlesim::Spawn>("/spawn");

	// init Spawn
	turtlesim::Spawn srv;
	srv.request.x = 2.0;
	srv.request.y = 2.0;
	srv.request.name = "turtle2";
	
	ROS_INFO("Call service to spwan turtle[x:%0.6f, y:%0.6f, name:%s]", srv.request.x, srv.request.y, srv.request.name.c_str());

	add_turtle.call(srv);

	// show result
	ROS_INFO("Spwan turtle successfully [name:%s]", srv.response.name.c_str());

	return 0;
}

File: catkin_ws/src/learning_service/scripts/turtle_spawn.py
#!/usr/bin/env python

import sys
import rospy
from turtlesim.srv import Spawn

def turtle_spawn():
    # ROS node init
    rospy.init_node("turtle_spawn")

    # wait for /spawn service, create a client, connect the /spawn service
    rospy.wait_for_service('/spawn')
    try:
        add_turtle = rospy.ServiceProxy('/spawn', Spawn)

        # send request
        response = add_turtle(2.0, 2.0, 0.0, "turtle2")
        return response.name
    except rospy.ServiceException, e:
        print "Service call failed: %s" % e

if __name__ == '__main__':
    # show result
    print "Spawn turtle successfully [name:%s]" % (turtle_spawn())
